lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -0,0 +1,689 @@
+        self._alert_rules_path = alert_rules_path
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_changed, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_departed, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_broken, self._handle_endpoints_changed)
+        self.framework.observe(on_relation.relation_joined, self._set_alerts_on_relation_changed)
+        self.framework.observe(self._charm.on.upgrade_charm, self._set_alerts_to_all_relation)
+
+    def _handle_endpoints_changed(self, event: RelationEvent):
+        self.on.endpoints_changed.emit(relation_id=event.relation.id)
+
+    def _set_alerts_on_relation_changed(self, event: RelationEvent):
+        self._set_alerts_to_relation(event.relation)
+
+    def _set_alerts_to_all_relation(self, _):
+        for relation in self.model.relations[self._relation_name]:
+            self._set_alerts_to_relation(relation)
+
+    def _set_alerts_to_relation(self, relation: Relation):
+        if alert_groups := self._labeled_alert_groups:
+            relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": alert_groups})
+
+    @property
+    def endpoints(self) -> List[Dict[str, str]]:
+        """A config object ready to be dropped in to a prometheus config file.
+
+        Returns:
+            A list of dictionaries where each dictionary provides information about
+            a single remote_write endpoint.
+        """
+        endpoints = []
+        for relation in self.model.relations[self._relation_name]:
+            for unit in relation.units:
+                if unit.app is self._charm.app:
+                    # This is a peer unit
+                    continue
+
+                if remote_write := relation.data[unit].get("remote_write"):
+                    deserialized_remote_write = json.loads(remote_write)
+                    endpoints.append(
+                        {
+                            "url": deserialized_remote_write["url"],
+                        }
+                    )
+
+        return endpoints
+
+    def _label_alert_topology(self, rule) -> dict:
+        """Insert juju topology labels into an alert rule.
+
+        Args:
+            rule: a dictionary representing a prometheus alert rule.
+
+        Returns:
+            a dictionary representing prometheus alert rule with juju
+            topology labels.
+        """
+        metadata = self._consumer_metadata
+        labels = rule.get("labels", {})
+        labels["juju_model"] = metadata["model"]
+        labels["juju_model_uuid"] = metadata["model_uuid"]
+        labels["juju_application"] = metadata["application"]
+        rule["labels"] = labels
+        return rule
+
+    def _label_alert_expression(self, rule) -> dict:
```

Comment:
iirc, this function and its neighbors were refactored quite a bit in `prometheus_scrape`.